File: seminar-09/kruskal/python/kruskal.py
class DSU:
    def __init__(self, size):
        self.parent = [i for i in range(size)]
        self.size = [1 for _ in range(size)]

    def get(self, x):
        if self.parent[x] == x:
            return x
        leader = self.get(self.parent[x])
        self.parent[x] = leader
        return leader

    def union(self, a, b):
        a = self.get(a)
        b = self.get(b)
        if a == b:
            return False

        if self.size[a] < self.size[b]:
            a, b = b, a

        self.parent[b] = a
        self.size[a] += self.size[b]
        return True


def kruskal(n, edges):
    """
    n — число вершин
    edges — список рёбер в формате (w, u, v)
    вершины нумеруются от 0 до n - 1
    """
    edges.sort()
    dsu = DSU(n)
    mst_weight = 0

    for w, u, v in edges:
        if dsu.union(u, v):
            mst_weight += w

    return mst_weight

File: seminar-09/kruskal/java/Kruskal.java
import java.util.Collections;
import java.util.List;

public class Kruskal {

    static class DSU {
        int[] parent;
        int[] size;

        DSU(int n) {
            parent = new int[n];
            size = new int[n];
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                size[i] = 1;
            }
        }

        int get(int x) {
            if (parent[x] == x) return x;
            return parent[x] = get(parent[x]);
        }

        boolean union(int a, int b) {
            a = get(a);
            b = get(b);
            if (a == b) return false;

            if (size[a] < size[b]) {
                int tmp = a;
                a = b;
                b = tmp;
            }

            parent[b] = a;
            size[a] += size[b];
            return true;
        }
    }

    static class Edge implements Comparable<Edge> {
        int u;
        int v;
        int w;

        Edge(int u, int v, int w) {
            this.u = u;
            this.v = v;
            this.w = w;
        }

        public int compareTo(Edge other) {
            return Integer.compare(this.w, other.w);
        }
    }

    public static long kruskal(int n, List<Edge> edges) {
        Collections.sort(edges);

        DSU dsu = new DSU(n);
        long mstWeight = 0;

        for (Edge e : edges) {
            if (dsu.union(e.u, e.v)) {
                mstWeight += e.w;
            }
        }

        return mstWeight;
    }
}
